modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/TargetRequest.java

```diff
@@ -145,29 +145,28 @@ protected long getContentLength(NHttpClientConnection conn) {
         return contentLength;
     }
     
-    protected HttpRequest getHttpRequest(NHttpClientConnection conn, String path) {
+    protected HttpRequest getHttpRequest(NHttpClientConnection conn, String path, boolean needToProcessChunking) {
         long contentLength = getContentLength(conn);
         
         MessageContext requestMsgCtx = TargetContext.get(conn).getRequestMsgCtx();
         
-        HttpRequest request = 
-                new BasicHttpEntityEnclosingRequest(method, path, version != null ? version : HttpVersion.HTTP_1_1);
+        HttpRequest request = null;
 
-        BasicHttpEntity entity = new BasicHttpEntity();
-        
-        ((BasicHttpEntityEnclosingRequest) request).setEntity(entity);
-        
         boolean forceContentLength = 
                 requestMsgCtx.isPropertyTrue(NhttpConstants.FORCE_HTTP_CONTENT_LENGTH);
         boolean forceContentLengthCopy = 
                 requestMsgCtx.isPropertyTrue(PassThroughConstants.COPY_CONTENT_LENGTH_FROM_INCOMING);
                             
-        if (forceContentLength) {
-            entity.setChunked(false);
+        if (forceContentLength && needToProcessChunking) {
             if (forceContentLengthCopy && contentLength != -1) {
                 if (log.isDebugEnabled()) {
                     log.debug("Set ContentLength : " + contentLength);
                 }
+                request = new BasicHttpEntityEnclosingRequest(method, path,
+                        version != null ? version : HttpVersion.HTTP_1_1);
+                BasicHttpEntity entity = new BasicHttpEntity();
+                ((BasicHttpEntityEnclosingRequest) request).setEntity(entity);
+                entity.setChunked(false);
                 entity.setContentLength(contentLength);
             } else if (!hasEntityBody) {
```

Comment:
Hi Kim,

I just noticed that there is a possible NPE caused due to my previous suggestion.
We need to add else clause for this if-elseif. Otherwise, if there is a request that does not match the criteria we check here, this will cause NPE. hense let's create BasicHttpEntityEnclosingRequest for as default in else block

Thanks,

Reply:
Hi Kim,

I will fix this. Rest of the fix looks ok.